### DISCUSSION TITLE
MobilityDB for storing huge number of realtime coordinates?

### DISCUSSION THREAD

**Discussion by bendecaf** [ULB-CoDE-WIT/MobilityDB#511] [General]
Hi,

I have one rather fundamental question on how to use MobilityDB properly and where its strengths are. 

Suppose you receive hundreds of vehicle coordinates per minute from phone apps. Can MobilityDB be used to store these coordinates as they arrive, for example, in a trajectory of the day for each verhicle or is MobilityDB more designed to read the coordinates after they have been stored for further analysis?

When reading the MobilityDB workshop document, it is my impression that the usage is to primarily read the data from an existing dataset (csv or other tables) and use the resulting trajectory to analyze it. Is this assumption correct, that it is not a good idea to store new coordinates directly in a trajectory as they arrive?

Thanks
Ben

**Comment by bendecaf** [comment]
Another question that came up for me: is there any real limit on how big, for example, a tgeompoint dataset realistically should become to be efficiently usable for analysis?

**Reply by mschoema** [reply]
This depends on what you mean by 'big'. There are two aspects that determine the size of a trajectory dataset.:

1. The number of trajectories
2. The (average) size of each individual trajectory, counted in number of segments.

Increasing the number of trajectories simply means increasing the number of tuples in your table. This will of course have an impact on the performance of your queries, but this is the same problem that any other large postgres table will have. In this case, indices on the `tgeompoint` can probably improve query performance.

On the other hand, increasing the size of each trajectory is a more complex matter. When working with long trajectories that span large spatio-temporal extents, indices become less performant and the functins and operators will have to be applied on more data. This is problematic mainly for complex functions that have a higher than linear complexity. When working with 'never-ending' data, such as streaming data, it is thus needed to sometimes split trajectories. A typical way is to collect trajectories at a hour, day or week granularity.

**Comment by mschoema** [comment]
Hi Ben,

MobilityDB is indeed primarily built for analysis of historical data. Updating an existing trajectory to add new incoming data points is possible, but requires an `UPDATE` query, which is rather expensive for large trajectories. It is thus not ideal to call such `UPDATE` queries frenquently, for example for each new data points. However, you can always try this method out first. It could still be fast enough for your use-case. If you discover that calling these frequent `UPDATE` queries it too slow, there are different ways to avoid this problem depending on your requirements. 

A first simple method would be to store the coordinates and timestamps of the last day or hour as individual tuples in a separate table, where they can be efficiently added using `INSERT` queries. Then, you could periodically collect this data into trajectories, for example using a [periodic job](https://github.com/citusdata/pg_cron). This would allow you to have efficient inserts of new coordinates and store your older data in `tgeompoint` format to allow for efficient analysis. The main downside is that you cannot easily call MobilityDB functions on the newer real-time data that has not yet been collected into trajectories.

In case you need to call MobilityDB functions on your real-time data, the best solution might be to use the [MEOS](https://libmeos.org/) C library or [PyMEOS](https://github.com/MobilityDB/PyMEOS) python library to process your data stream in-memory and then periodically send collected trajectories to a MobilityDB database for persistent storage.

I hope this answers your questions.
If you have questions related to a specific use-case or scenario, we can also discuss this futher to help you find a good solution.

Best,
Maxime

**Reply by bendecaf** [reply]
Thank you very much for the quick reply. I'll have a look at MEOS (libmeos.org seems to have a certificate problem at the moment). My current intuition going forward is your suggestion to store coordinates with timestamp in a separate timescaledb hypertable and create the trajectories as necessary for analysis. I'll do a few tests how well that process works with an artifically filled database.

**Comment by estebanzimanyi** [comment]
This question was answered above: MobilityDB is primarily built for analysis of historical trajectories; for high-rate ingestion the recommended pattern is to buffer incoming points as plain rows and periodically assemble them into temporal values, rather than issuing frequent UPDATEs on a large trajectory. Closing as answered; further ingestion questions are welcome as a new discussion.
